### PR TITLE
[CARBONDATA-1822][Spark-Integration] Support DDL to register the CarbonData table from existing carbon table data

### PR DESCRIPTION
**Pull request by mohammadshahidkhan**
**Problem:**
Support DDL command to create the table from existing table data. 
**Solution:**
Existing spark refresh table DDL command will be overridden in the carbon to support the register
carbon table with hivemetastore.
  **REFRESH TABLE $dbName.$tbName;** 
  __
   _**i.  For non carbon table the refresh table request will be passed to the spark.**_
  ii.  The database path will be retrieved from hive catalog & the database path will be scanned to get 
       table. 
 iii.  The table schema will be read to get columns details. 
 iv.  The table will be registered to the hive catalog with below details.
 
_CREATE TABLE $tbName USING {filed1 DataType, field2 DataType... } carbondata OPTIONS (tableName "$tbName", 
dbName "$dbName", 
tablePath "$tablePath",)_ 

**Precondition**: 
    i. The user has to create the database and Before executing this command the old table schema and 
       data should be copied into the database location. 
   ii. The table should be copied at the database location.
  iii. If the table is aggregate table then all the aggregate tables should  be copied to the  in database location . 
  iv. The carbon table should be from the compatible carbon version.
**Validation**: 
   i. If database does not exist then the registration will fail. 
   ii. The table will be registered only if same table name is not already  registered. 
  iii. If the table contains the aggregate tables then all the aggregate  tables should be registered to hive 
       catalog and if any of the aggregate table does not exist then the table creation operation should fail. 

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [X] Any interfaces changed?
     Overridden the spark refresh table command.

 - [X] Any backward compatibility impacted?
    No
 - [X] Document update required?
    Yes
 - [X] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       Yes 
       Added new test case to validate the register of normal table and aggregate table.
 - [X] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1534/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1938/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1545/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1946/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1550/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1551/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1950/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1560/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1562/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1959/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1961/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1595/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/1991/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1605/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1608/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2000/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2002/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1614/

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/RefreshCarbonTableCommand.scala [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+    var locks = List.empty[ICarbonLock]
+    try {
+      val tablePath = databaseLocation + CarbonCommonConstants.FILE_SEPARATOR + tableName
+      val absoluteTableIdentifier = AbsoluteTableIdentifier.from(tablePath, databaseName, tableName)
+      locks = CommonUtil.acquireLock(locksToBeAcquired, absoluteTableIdentifier)
+      // 2.1 check if the table already register with hive then ignore and continue with the next
+      // schema
+      if (!sparkSession.sessionState.catalog.listTables(databaseName)
+        .exists(_.table.equalsIgnoreCase(tableName))) {
+        val carbonTablePath = CarbonStorePath.getCarbonTablePath(absoluteTableIdentifier)
```
Comment (ravipesala) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/RefreshCarbonTableCommand.scala]: Please move reading of TableInfo using tablePath to the CarbonMetaStore

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/RefreshCarbonTableCommand.scala [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+    // Steps
+    // 1. get table path
+    // 2. perform the below steps
+    // 2.1 check if the table already register with hive then ignore and continue with the next
+    // schema
+    // 2.2 register the table with the hive check if the table being registered has aggregate table
+    // then do the below steps
+    // 2.2.1 validate that all the aggregate tables are copied at the store location.
+    // 2.2.2 Register the aggregate tables
+    val locksToBeAcquired = List(LockUsage.METADATA_LOCK, LockUsage.DROP_TABLE_LOCK)
```
Comment (ravipesala) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/RefreshCarbonTableCommand.scala]: why locks are required to take here? we are not updating any files right, we are just updating to DB, I think concurrent scenarios it can take care internally.

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/RefreshCarbonTableCommand.scala [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+      val fields = new Array[String](columns.size)
+      columns.map(column => fields(column.getSchemaOrdinal)
+        = column.getColumnName + ' ' + prepareDataType(column))
+      val operationContext = new OperationContext
+      val createTablePreExecutionEvent: RefreshTablePreExecutionEvent =
+        new RefreshTablePreExecutionEvent(sparkSession,
+          tableInfo.getOrCreateAbsoluteTableIdentifier().getCarbonTableIdentifier,
+          tablePath)
+      OperationListenerBus.getInstance.fireEvent(createTablePreExecutionEvent, operationContext)
+      sparkSession.sql(
```
Comment (ravipesala) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/RefreshCarbonTableCommand.scala]: I think you  can directly CreateTableCOmmand here instead of creating sql here.

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2008/

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/strategy/DDLStrategy.scala [hunk 4]
```diff
@@ -193,6 +197,32 @@ class DDLStrategy(sparkSession: SparkSession) extends SparkStrategy {
         ExecutedCommandExec(
           CarbonAlterTableUnsetCommand(tableName, propKeys, ifExists, isView)) :: Nil
       }
+      case RefreshTable(tableIdentifier) =>
+        val isCarbonTable = CarbonEnv.getInstance(sparkSession).carbonMetastore
```
Comment (ravipesala) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/strategy/DDLStrategy.scala]: here just call `RefreshCarbonTableCommand ` and at the end of that command you can call `RefreshTable`

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/RefreshCarbonTableCommand.scala (hunk 3 above)
Comment (mohammadshahidkhan) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/RefreshCarbonTableCommand.scala]: @ravipesala Please elaborate how to use the same i don't see any instance.

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/strategy/DDLStrategy.scala (hunk 4 above)
Comment (mohammadshahidkhan) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/strategy/DDLStrategy.scala]: Fixed

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/RefreshCarbonTableCommand.scala (hunk 1 above)
Comment (mohammadshahidkhan) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/RefreshCarbonTableCommand.scala]: moved to CarbonMetaStore.getWrapperTableInfo(absoluteTableIdentifier)

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/RefreshCarbonTableCommand.scala (hunk 2 above)
Comment (mohammadshahidkhan) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/RefreshCarbonTableCommand.scala]: removed

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1625/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2016/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1647/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2036/

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/RefreshCarbonTableCommand.scala [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+        // 2.2.1 validate that all the aggregate tables are copied at the store location.
+        val dataMapSchemaList = tableInfo.getDataMapSchemaList
+        if (null != dataMapSchemaList && dataMapSchemaList.size() != 0) {
+          // validate all the aggregate tables are copied at the storeLocation
+          val allExists = validateAllAggregateTablePresent(databaseName,
+            dataMapSchemaList, databaseLocation)
+          if (!allExists) {
+            // fail the register operation
+            val msg = s"Table registration with Database name [$databaseName] and Table name " +
+                      s"[$tableName] failed. All the aggregate Tables for table [$tableName] is" +
```
Comment (ravipesala) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/RefreshCarbonTableCommand.scala]: Better print the agg table names which are not copied

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/413/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2070/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1685/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/418/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2076/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1690/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/439/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2091/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2092/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/441/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1709/

**Comment by ravipesala** [comment]
retest sdv please

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1711/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2096/

**Comment by mohammadshahidkhan** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/451/

**Comment by ravipesala** [comment]
retest this please

File: integration/spark-common/src/main/scala/org/apache/carbondata/spark/util/CommonUtil.scala [hunk 6]
```diff
@@ -876,5 +876,4 @@ object CommonUtil {
       }
     }
   }
-
```
Comment (ravipesala) [inline integration/spark-common/src/main/scala/org/apache/carbondata/spark/util/CommonUtil.scala]: remove unnecessary change

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/452/

File: integration/spark2/src/main/scala/org/apache/carbondata/spark/util/CarbonSparkUtil.scala [hunk 7]
```diff
@@ -51,4 +55,90 @@ object CarbonSparkUtil {
...
+   * @param columnName
+   * @param dataType
+   * @param aggregateType
+   * @param parentColumnId
+   * @param parentTableName
+   * @param parentDatabaseName
+   * @param parentTableId
+   * @return fields object
+   */
+  def getField(columnName: String,
```
Comment (ravipesala) [inline integration/spark2/src/main/scala/org/apache/carbondata/spark/util/CarbonSparkUtil.scala]: Why do you require this code? you ultimately require `rawSchema` , that you can anyway get from `dataType.catalogString`, just append the comment to it and use in createtablecommand

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/table/CarbonCreateTableCommand.scala [hunk 8]
```diff
@@ -81,12 +90,10 @@ case class CarbonCreateTableCommand(
...
     if (createDSTable) {
       try {
-        val fields = new Array[Field](cm.dimCols.size + cm.msrCols.size)
-        cm.dimCols.foreach(f => fields(f.schemaOrdinal) = f)
-        cm.msrCols.foreach(f => fields(f.schemaOrdinal) = f)
-
-        sparkSession.sparkContext.setLocalProperty(EXECUTION_ID_KEY, null)
         val tablePath = tableIdentifier.getTablePath
+        val carbonRelation = CarbonSparkUtil.createCarbonRelation(tableInfo, tablePath)
+        val fields: Array[Field] = CarbonSparkUtil.getColumnSchemaFields(carbonRelation)
```
Comment (ravipesala) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/table/CarbonCreateTableCommand.scala]: you don't require fields here, you just require rawSchema

File: integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonFileMetastore.scala [hunk 9]
```diff
@@ -523,4 +523,25 @@ class CarbonFileMetastore extends CarbonMetaStore {
...
   }
+
+  /**
+   * the method returns the Wrapper TableInfo
+   *
+   * @param absoluteTableIdentifier
+   * @param sparkSession
+   * @return
+   */
+  override def getWrapperTableInfo(absoluteTableIdentifier: AbsoluteTableIdentifier)
```
Comment (ravipesala) [inline integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonFileMetastore.scala]: Just name name as `getTableInfo`

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1717/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1721/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/457/

**Comment by mohammadshahidkhan** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/465/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1726/

File: integration/spark-common/src/main/scala/org/apache/carbondata/events/RefreshTableEvents.scala [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+ */
+
+package org.apache.carbondata.events
+
+import org.apache.spark.sql._
+
+import org.apache.carbondata.core.metadata.AbsoluteTableIdentifier
+
+/**
+ * Class for handling operations before start of a load process.
```
Comment (jackylk) [inline integration/spark-common/src/main/scala/org/apache/carbondata/events/RefreshTableEvents.scala]: There already event to load process

File: integration/spark-common/src/main/scala/org/apache/spark/sql/catalyst/CarbonDDLSqlParser.scala [hunk 11]
```diff
@@ -1044,48 +1045,48 @@ abstract class CarbonDDLSqlParser extends AbstractCarbonSparkSQLParser {
...
       case "long" => Field(field.column, Some("Long"), field.name, Some(null), field.parent,
-        field.storeType, field.schemaOrdinal, field.precision, field.scale, field.rawSchema
-      )
+        field.storeType, field.schemaOrdinal, field.precision, field.scale, field.rawSchema,
+        field.columnComment)
       case "double" => Field(field.column, Some("Double"), field.name, Some(null), field.parent,
-        field.storeType, field.schemaOrdinal, field.precision, field.scale, field.rawSchema
-      )
+        field.storeType, field.schemaOrdinal, field.precision, field.scale, field.rawSchema,
+        field.columnComment)
```
Comment (jackylk) [inline integration/spark-common/src/main/scala/org/apache/spark/sql/catalyst/CarbonDDLSqlParser.scala]: Is this change for this PR? If not for this PR, please raise another PR for it

File: integration/spark-common/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchemaCommon.scala [hunk 12]
```diff
@@ -47,6 +47,7 @@ import org.apache.carbondata.spark.util.DataTypeConverterUtil
 
 case class TableModel(
     ifNotExistsSet: Boolean,
+    var databaseName: String,
```
Comment (jackylk) [inline integration/spark-common/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchemaCommon.scala]: There is already a databaseNameOp in line 51

File: integration/spark-common/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchemaCommon.scala [hunk 13]
```diff
@@ -320,13 +322,8 @@ class AlterTableColumnSchemaGenerator(
...
 object TableNewProcessor {
   def apply(
-      cm: TableModel,
-      identifier: AbsoluteTableIdentifier): TableInfo = {
-    new TableNewProcessor(
-      cm,
-      identifier.getDatabaseName,
-      identifier.getTableName,
-      identifier.getTablePath).process
+      cm: TableModel): TableInfo = {
```
Comment (jackylk) [inline integration/spark-common/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchemaCommon.scala]: Do not change this, otherwise tablePath will be lost

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/RefreshCarbonTableCommand.scala [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+    // Steps
+    // 1. get table path
+    // 2. perform the below steps
+    // 2.1 check if the table already register with hive then ignore and continue with the next
+    // schema
+    // 2.2 register the table with the hive check if the table being registered has aggregate table
+    // then do the below steps
+    // 2.2.1 validate that all the aggregate tables are copied at the store location.
+    // 2.2.2 Register the aggregate tables
+    val tablePath = databaseLocation + CarbonCommonConstants.FILE_SEPARATOR + tableName
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/RefreshCarbonTableCommand.scala]: use `CarbonEnv.getTablePath` to get the path

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/table/CarbonCreateTableCommand.scala [hunk 15]
```diff
@@ -29,25 +29,34 @@ import org.apache.carbondata.common.logging.LogServiceFactory
...
 import org.apache.carbondata.core.metadata.AbsoluteTableIdentifier
-import org.apache.carbondata.core.metadata.schema.table.TableInfo
+import org.apache.carbondata.core.metadata.schema.table.{CarbonTable, TableInfo}
 import org.apache.carbondata.core.util.CarbonUtil
 import org.apache.carbondata.events.{CreateTablePostExecutionEvent, CreateTablePreExecutionEvent, OperationContext, OperationListenerBus}
+import org.apache.carbondata.spark.util.CarbonSparkUtil
 
 case class CarbonCreateTableCommand(
-    cm: TableModel,
+    tableInfo: TableInfo,
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/table/CarbonCreateTableCommand.scala]: I do not think you can create the TableInfo in parser, since it does not have the sparkSession, so the dbName maybe wrong. You can need to use `CarbonEnv.getDatabaseName` to get the dbName, it requires sparkSession

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/table/CarbonCreateTableCommand.scala [hunk 16]
```diff
@@ -81,15 +90,13 @@ case class CarbonCreateTableCommand(
...
     if (createDSTable) {
       try {
-        val fields = new Array[Field](cm.dimCols.size + cm.msrCols.size)
-        cm.dimCols.foreach(f => fields(f.schemaOrdinal) = f)
-        cm.msrCols.foreach(f => fields(f.schemaOrdinal) = f)
-
-        sparkSession.sparkContext.setLocalProperty(EXECUTION_ID_KEY, null)
         val tablePath = tableIdentifier.getTablePath
+        val carbonRelation = CarbonSparkUtil.createCarbonRelation(tableInfo, tablePath)
+        val rawSchema = CarbonSparkUtil.getRawSchema(carbonRelation)
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/table/CarbonCreateTableCommand.scala]: Why this is needed

File: integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonMetaStore.scala [hunk 17]
```diff
@@ -142,6 +142,10 @@ trait CarbonMetaStore {
 
   def getTableFromMetadataCache(database: String, tableName: String): Option[CarbonTable]
 
+  def getTableInfo(absoluteTableIdentifier: AbsoluteTableIdentifier)
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonMetaStore.scala]: Why this is required? It is better not to add more interface

**Comment by jackylk** [comment]
It seems there is a lot of change related to column comment, please separate them to another PR if not  related to this PR

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/480/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2108/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1732/

File: integration/spark-common/src/main/scala/org/apache/spark/sql/catalyst/CarbonDDLSqlParser.scala (hunk 11 above)
Comment (mohammadshahidkhan) [reply integration/spark-common/src/main/scala/org/apache/spark/sql/catalyst/CarbonDDLSqlParser.scala]: this is needed as part of this pr only

File: integration/spark-common/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchemaCommon.scala (hunk 13 above)
Comment (mohammadshahidkhan) [reply integration/spark-common/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchemaCommon.scala]: We have to pass the TableInfo to the CarbonCreateTableCommand so that we can use the 
CarbonCreateTableCommand  flow from the RefreshCarbonTableCommand.
Any ways for supporting the external table path, the tablePath is being passed to CarbonCreateTableCommand

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/RefreshCarbonTableCommand.scala (hunk 14 above)
Comment (mohammadshahidkhan) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/RefreshCarbonTableCommand.scala]: fixed

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/table/CarbonCreateTableCommand.scala (hunk 15 above)
Comment (mohammadshahidkhan) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/table/CarbonCreateTableCommand.scala]: correct if user does not pass the database name while creating table, then database name will not 
there that is why added the following code in CarbonCreateTableCommand.processMetadata
  var databaseOpt : Option[String] = None
    if(tableInfo.getDatabaseName != null) {
      databaseOpt = Some(tableInfo.getDatabaseName)
    }
    val dbName = CarbonEnv.getDatabaseName(databaseOpt)(sparkSession)

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/table/CarbonCreateTableCommand.scala (hunk 16 above)
Comment (mohammadshahidkhan) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/table/CarbonCreateTableCommand.scala]: While registering the table with hive we need rawSchema,
earlier this we were getting from the column  Field

File: integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonMetaStore.scala (hunk 17 above)
Comment (mohammadshahidkhan) [reply integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonMetaStore.scala]: many places we are reading schema from the tablePath, ie we decided to move it at common place.
Once this PR is merged will refactor code from other places to use this interface.

File: integration/spark-common/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchemaCommon.scala (hunk 12 above)
Comment (mohammadshahidkhan) [reply integration/spark-common/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchemaCommon.scala]: fixed

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/481/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1742/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1745/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2128/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2129/

File: integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonMetaStore.scala (hunk 17 above)
Comment (jackylk) [reply integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonMetaStore.scala]: This idea is good, but I think you do not need to put this in CarbonMetaStore trait, right? Reading schema from tablePath can be an utility.

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/table/CarbonCreateTableCommand.scala (hunk 15 above)
Comment (jackylk) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/table/CarbonCreateTableCommand.scala]: This approach is fine. But why you need to modify it in first place? TableModel can't work?

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/RefreshCarbonTableCommand.scala [hunk 18]
```diff
@@ -0,0 +1,210 @@
...
+import org.apache.carbondata.core.metadata.schema.table.column.ColumnSchema
+import org.apache.carbondata.core.util.CarbonProperties
+import org.apache.carbondata.core.util.path.CarbonStorePath
+import org.apache.carbondata.events.{OperationContext, OperationListenerBus, RefreshTablePostExecutionEvent, RefreshTablePreExecutionEvent}
+
+/**
+ * Command to register carbon table from existing carbon table data
+ */
+case class RefreshCarbonTableCommand(
+    dbName: Option[String],
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/RefreshCarbonTableCommand.scala]: change to databaseNameOp, make it consistent with other command

File: integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonMetaStore.scala (hunk 17 above)
Comment (mohammadshahidkhan) [reply integration/spark2/src/main/scala/org/apache/spark/sql/hive/CarbonMetaStore.scala]: fixed added in schema redear

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/RefreshCarbonTableCommand.scala (hunk 18 above)
Comment (mohammadshahidkhan) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/RefreshCarbonTableCommand.scala]: fixed change to databaseNameOp

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/554/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/555/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/557/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/561/

**Comment by mohammadshahidkhan** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/564/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/1811/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2183/

**Comment by ravipesala** [comment]
LGTM
